feat(validation): include configurability info in rule definition

f1ames opened the pull request:
This PR fixes https://github.com/kubeshop/monokle-core/issues/397.

## Changes

- Introduced `configurability` field in rule definition to indicate for 3rd-parties that rule is configurable. This is required for any tools integrating with the library to  reason about rules and provide correct UI.

I'm not fully convinced about naming, we have `defaultConfiguration`, `configuration` and now `configurability`. Also `configruability` is a definition of how rule can be configured while values are passed via `configuration.parameters`, not sure if this shouldn't be merged somehow.

## Fixes

- None.

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

f1ames commented:
> I'm not fully convinced about naming, we have `defaultConfiguration`, `configuration` and now `configurability`. Also `configruability` is a definition of how rule can be configured while values are passed via `configuration.parameters`, not sure if this shouldn't be merged somehow.

I already answered my question in https://github.com/kubeshop/monokle-core/pull/402/commits/4dcb375552963bdafb6cbb49c57657ad0d037f1e 😄 While for the simplified rule initialization (`RuleInit`) we need default value, in the full SARIF definition (`RuleMetadata`) it's not needed (even it shouldn't be there, since it's quite confusing), because we have `defaultConfiguration.parameters` - and this prop should held default values.

f1ames commented:
Ready for review as it was tested with https://github.com/kubeshop/monokle-saas/pull/1464.

WitoDelnat commented on packages/validation/src/validators/custom/config.ts:
```diff
@@ -1,5 +1,5 @@
 import type {Document, ParsedNode} from 'yaml';
-import {RuleLevel, reportingDescriptorRelationship} from '../../node';
+import {RuleConfigurability, RuleLevel, reportingDescriptorRelationship} from '../../node';
```
node.js'; 

iirc, imports must end with .js normally, not sure why this is working as it already was like this 🤔 Maybe because it area only types?

f1ames replied:
Fixed 👍

f1ames commented on packages/validation/src/common/sarif.ts:
```diff
@@ -114,6 +114,8 @@ export type RuleMetadata<TProperties = {}> = {
 
   relationships?: reportingDescriptorRelationship[];
+
+  configurability?: RuleConfigurability;
```
Ahhh, right. For a moment I though about SARIF schema... 😅 The plugin wasn't enabled in the test, that's why I overlooked it. I will take a look on specs, and probably end up with custom properties as you suggested 👍

f1ames commented:
@WitoDelnat when I fixed https://github.com/kubeshop/monokle-core/pull/402#discussion_r1235174450, then SARIF schema validation test started complaining about `parameters` type. After checking SARIF specs, it turned out it's expected to be `Property bag` (see https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Ref508894764 and https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Ref493408960). And the test did not detect it before because it wasn't used in any rules there.

Before we stored rule config like `parameters: string | number | any[]` but since it needs to be key-value type I reworked it to comply with specs. Reworked related types and code. Now, rule config is stored as `.parameters.configValue: any`.

And this also opens a door to store other custom props in `parameters` object (kind of, what I already did with `name` for dynamic labels).

Done in https://github.com/kubeshop/monokle-core/pull/402/commits/ffbc065cbecc7205a4e530303adabadd305512d4.

WitoDelnat reviewed (approved):
LGTM ✅
